src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -7130,24 +7144,188 @@ public override IOperation VisitArgument(IArgumentOperation operation, int? capt
             throw ExceptionUtilities.Unreachable;
         }
 
-        public override IOperation VisitUsingDeclaration(IUsingDeclarationOperation operation, int? argument)
+        public override IOperation VisitUsingDeclaration(IUsingDeclarationOperation operation, int? captureIdForResult)
         {
             throw ExceptionUtilities.Unreachable;
         }
 
-        public override IOperation VisitWith(IWithOperation operation, int? argument)
+        public override IOperation VisitWith(IWithOperation operation, int? captureIdForResult)
         {
+            if (operation.Type!.IsAnonymousType)
+            {
+                return handleAnonymousTypeWithExpression((WithOperation)operation, captureIdForResult);
+            }
+
             EvalStackFrame frame = PushStackFrame();
             // Initializer is removed from the tree and turned into a series of statements that assign to the cloned instance
             IOperation visitedInstance = VisitRequired(operation.Operand);
+            IOperation cloned;
+            if (operation.Type.IsValueType)
+            {
+                cloned = visitedInstance;
+            }
+            else
+            {
+                cloned = operation.CloneMethod is null
+                    ? MakeInvalidOperation(visitedInstance.Type, visitedInstance)
+                    : new InvocationOperation(operation.CloneMethod, visitedInstance,
+                        isVirtual: true, arguments: ImmutableArray<IArgumentOperation>.Empty,
+                        semanticModel: null, operation.Syntax, operation.Type, isImplicit: true);
+            }
+
+            return PopStackFrame(frame, HandleObjectOrCollectionInitializer(operation.Initializer, cloned));
+
+            // For `old with { Property = ... }` we're going to do the same as `new { Property = ..., OtherProperty = old.OtherProperty }`
+            IOperation handleAnonymousTypeWithExpression(WithOperation operation, int? captureIdForResult)
+            {
+                Debug.Assert(operation.Type!.IsAnonymousType);
+                SpillEvalStack(); // before entering a new region, we ensure that anything that needs spilling was spilled
+
+                // The outer region holds captures for all the values for the anonymous object creation
+                var outerCaptureRegion = CurrentRegionRequired;
+
+                // The inner region holds the capture for the operand (ie. old value)
+                var innerCaptureRegion = new RegionBuilder(ControlFlowRegionKind.LocalLifetime);
+                EnterRegion(innerCaptureRegion);
+
+                var initializers = operation.Initializer.Initializers;
+
+                var properties = operation.Type.GetMembers()
```

Comment:
>properties [](http://example.com/codeflow?start=20&length=10)

Consider if we could avoid allocating this array. It feels like we could simply iterate through GetMembers array twice, when we need to.